From 334f1af51368f02c36a85877af46850cbb0f8d12 Mon Sep 17 00:00:00 2001
From: P Praneesh <quic_ppranees@quicinc.com>
Date: Tue, 15 Nov 2022 19:02:52 +0530
Subject: [PATCH] mac80211: fix RCU stall while updating mesh PS STA status

While mesh node sleeps during powersave mode, mac80211 buffers frames
until it takes up, then the frames are transmitted. But
ieee80211_sta_ps_deliver_wakeup holds the sta->ps_lock and transmits
the pending skbs by scheduling the tasklet, but it again takes the
sta->ps_lock which results in RCU stall.

Fix it by disabling bottom half before calling
ieee80211_sta_ps_deliver_wakeup to prevent tasklet schedule.
Other callers already disabled bottom half before calling this
function.

[30778.398103] rcu: INFO: rcu_preempt self-detected stall on CPU
[30778.398121] rcu: 	2-....: (1 GPs behind) idle=bde/1/0x4000000000000004 softirq=1149236/1149236 fqs=1047
[30778.402812] 	(t=2100 jiffies g=3339929 q=112)
[30778.412357] Task dump for CPU 2:
[30778.416608] wpa_supplicant  R  running task        0 31738      1 0x00000002
[30778.419911] Call trace:
[30778.426943]  dump_backtrace+0x0/0x15c
[30778.429109]  show_stack+0x14/0x1c
[30778.432930]  sched_show_task+0x104/0x134
[30778.436228]  dump_cpu_task+0x40/0x274
[30778.440222]  rcu_dump_cpu_stacks+0x7c/0xd4
[30778.443780]  rcu_sched_clock_irq+0x350/0x824
[30778.447773]  update_process_times+0x2c/0x50
[30778.452200]  tick_sched_handle.isra.4+0x3c/0x44
[30778.456106]  tick_sched_timer+0x48/0x88
[30778.460619]  __hrtimer_run_queues+0xa0/0x140
[30778.464440]  hrtimer_interrupt+0xe4/0x214
[30778.468955]  arch_timer_handler_virt+0x28/0x3c
[30778.472861]  handle_percpu_devid_irq+0x84/0x12c
[30778.477200]  generic_handle_irq+0x18/0x2c
[30778.481626]  __handle_domain_irq+0x84/0xac
[30778.485793]  gic_handle_irq+0x74/0xbc
[30778.489785]  el1_irq+0xf0/0x1c0
[30778.493519]  queued_spin_lock_slowpath+0x98/0x2c0
[30778.496482]  ieee80211_beacon_cntdwn_is_complete+0x3b4/0x9c4 [mac80211]
[30778.501339]  ieee80211_schedule_txq+0x910/0x980 [mac80211]
[30778.507762]  ieee80211_tx_pending+0xf4/0x254 [mac80211]
[30778.513312]  tasklet_action_common.isra.2+0xcc/0x160
[30778.518433]  tasklet_action+0x24/0x2c
[30778.523641]  __do_softirq+0x10c/0x244
[30778.527198]  irq_exit+0x64/0xb4
[30778.530845]  __handle_domain_irq+0x88/0xac
[30778.533796]  gic_handle_irq+0x74/0xbc
[30778.537962]  el1_irq+0xf0/0x1c0
[30778.541697]  _raw_spin_unlock_irqrestore+0x14/0x40
[30778.544653]  ieee80211_add_pending_skbs+0x9c/0x108 [mac80211]
[30778.549514]  ieee80211_sta_ps_deliver_wakeup+0x2d0/0x3c0 [mac80211]
[30778.555331]  ieee80211_mps_sta_status_update+0xdc/0xe8 [mac80211]
[30778.561407]  ieee80211_channel_switch+0xde4/0x16b4 [mac80211]
[30778.567660]  ieee80211_channel_switch+0x11b8/0x16b4 [mac80211]
[30778.573386]  __cfg80211_send_event_skb+0x1f04/0x2a98 [cfg80211]
[30778.579113]  genl_family_rcv_msg+0x314/0x3b0
[30778.584925]  genl_rcv_msg+0x58/0x84
[30778.589438]  netlink_rcv_skb+0x94/0x10c
[30778.592650]  genl_rcv+0x34/0x48
[30778.596469]  netlink_unicast+0x138/0x1f8
[30778.599595]  netlink_sendmsg+0x28c/0x328
[30778.603763]  sock_sendmsg+0x18/0x2c
[30778.607667]  ____sys_sendmsg+0x2c4/0x350
[30778.610880]  ___sys_sendmsg+0x7c/0xc4
[30778.615046]  __sys_sendmsg+0x64/0xac
[30778.618605]  __arm64_sys_sendmsg+0x1c/0x24
[30778.622253]  el0_svc_common.constprop.0+0x98/0x114
[30778.626159]  el0_svc_handler+0x18/0x20
[30778.630933]  el0_svc+0x8/0x1b4

Signed-off-by: P Praneesh <quic_ppranees@quicinc.com>
---
 net/mac80211/mesh_ps.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/net/mac80211/mesh_ps.c b/net/mac80211/mesh_ps.c
index 3fbd0b9..b43a8aa 100644
--- a/net/mac80211/mesh_ps.c
+++ b/net/mac80211/mesh_ps.c
@@ -267,7 +267,9 @@ void ieee80211_mps_sta_status_update(struct sta_info *sta)
 		mps_dbg(sta->sdata, "start PS buffering frames towards %pM\n",
 			sta->sta.addr);
 	} else {
+		local_bh_disable();
 		ieee80211_sta_ps_deliver_wakeup(sta);
+		local_bh_enable();
 	}
 }
 
-- 
2.17.1

